// expo-cli/build/commands/publish.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.action = undefined;

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var action = exports.action = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(projectDir) {
    var options = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

    var channelRe, hasOptimized, nonInteractive, _ref2, allowOptimization, status, startedOurOwn, startOpts, recipient, _ref3, sdkVersion, buildStatus, _ref4, exp, result, url;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            channelRe = new RegExp(/^[a-z\d][a-z\d._-]*$/);

            if (options.releaseChannel && !channelRe.test(options.releaseChannel)) {
              (_log || _load_log()).default.error('Release channel name can only contain lowercase letters, numbers and special characters . _ and -');
              process.exit(1);
            }
            hasOptimized = _fs.default.existsSync(_path.default.join(projectDir, '/.expo-shared/assets.json'));
            nonInteractive = options.parent && options.parent.nonInteractive;

            if (!(!hasOptimized && !nonInteractive)) {
              _context.next = 13;
              break;
            }

            (_log || _load_log()).default.warn('It seems your assets have not been optimized yet.');
            _context.next = 8;
            return (0, (_prompt || _load_prompt()).default)({
              type: 'confirm',
              name: 'allowOptimization',
              message: 'Do you want to optimize assets now?'
            });

          case 8:
            _ref2 = _context.sent;
            allowOptimization = _ref2.allowOptimization;

            if (!allowOptimization) {
              _context.next = 13;
              break;
            }

            _context.next = 13;
            return (0, (_optimize || _load_optimize()).action)(projectDir);

          case 13:
            _context.next = 15;
            return (_xdl || _load_xdl()).Project.currentStatus(projectDir);

          case 15:
            status = _context.sent;
            startedOurOwn = false;

            if (!(status !== 'running')) {
              _context.next = 25;
              break;
            }

            (0, (_log || _load_log()).default)('Unable to find an existing Expo CLI instance for this directory, starting a new one...');
            (0, (_exit || _load_exit()).installExitHooks)(projectDir);

            startOpts = { reset: options.clear, nonPersistent: true };

            if (options.maxWorkers) {
              startOpts.maxWorkers = options.maxWorkers;
            }

            _context.next = 24;
            return (_xdl || _load_xdl()).Project.startAsync(projectDir, startOpts, !options.quiet);

          case 24:
            startedOurOwn = true;

          case 25:
            _context.next = 27;
            return (_sendTo || _load_sendTo()).default.getRecipient(options.sendTo);

          case 27:
            recipient = _context.sent;

            (0, (_log || _load_log()).default)('Publishing to channel \'' + options.releaseChannel + '\'...');

            _context.next = 31;
            return (_xdl || _load_xdl()).Exp.getPublishInfoAsync(projectDir);

          case 31:
            _ref3 = _context.sent;
            sdkVersion = _ref3.args.sdkVersion;
            _context.next = 35;
            return (_xdl || _load_xdl()).Project.buildAsync(projectDir, {
              mode: 'status',
              platform: 'all',
              current: true,
              releaseChannel: options.releaseChannel,
              sdkVersion: sdkVersion
            });

          case 35:
            buildStatus = _context.sent;
            _context.next = 38;
            return (_xdl || _load_xdl()).ProjectUtils.readConfigJsonAsync(projectDir);

          case 38:
            _ref4 = _context.sent;
            exp = _ref4.exp;


            if (buildStatus.userHasBuiltExperienceBefore && !buildStatus.userHasBuiltAppBefore && !options.duringBuild && !exp.isDetached) {
              (_log || _load_log()).default.warn('We noticed you did not build a standalone app with this SDK version and release channel before. ' + 'Remember that OTA updates will not work with the app built with different SDK version and/or release channel. ' + 'Read more: https://docs.expo.io/versions/latest/workflow/publishing/#limitations');
            }

            if (options.quiet) {
              (_simpleSpinner || _load_simpleSpinner()).default.start();
            }

            result = void 0;
            _context.prev = 43;
            _context.next = 46;
            return (_xdl || _load_xdl()).Project.publishAsync(projectDir, {
              releaseChannel: options.releaseChannel
            });

          case 46:
            result = _context.sent;
            url = result.url;


            if (options.quiet) {
              (_simpleSpinner || _load_simpleSpinner()).default.stop();
            }

            (0, (_log || _load_log()).default)('Published');
            (0, (_log || _load_log()).default)('Your URL is\n\n' + (_chalk || _load_chalk()).default.underline(url) + '\n');
            (_log || _load_log()).default.raw(url);

            if (!recipient) {
              _context.next = 55;
              break;
            }

            _context.next = 55;
            return (_sendTo || _load_sendTo()).default.sendUrlAsync(url, recipient);

          case 55:
            _context.prev = 55;

            if (!startedOurOwn) {
              _context.next = 59;
              break;
            }

            _context.next = 59;
            return (_xdl || _load_xdl()).Project.stopAsync(projectDir);

          case 59:
            return _context.finish(55);

          case 60:
            return _context.abrupt('return', result);

          case 61:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this, [[43,, 55, 60]]);
  }));

  return function action(_x) {
    return _ref.apply(this, arguments);
  };
}();

var _fs = _interopRequireDefault(require('fs'));

var _path = _interopRequireDefault(require('path'));

var _chalk;

function _load_chalk() {
  return _chalk = _interopRequireDefault(require('chalk'));
}

var _simpleSpinner;

function _load_simpleSpinner() {
  return _simpleSpinner = _interopRequireDefault(require('@expo/simple-spinner'));
}

var _xdl;

function _load_xdl() {
  return _xdl = require('xdl');
}

var _log;

function _load_log() {
  return _log = _interopRequireDefault(require('../log'));
}

var _prompt;

function _load_prompt() {
  return _prompt = _interopRequireDefault(require('../prompt'));
}

var _sendTo;

function _load_sendTo() {
  return _sendTo = _interopRequireDefault(require('../sendTo'));
}

var _optimize;

function _load_optimize() {
  return _optimize = require('./optimize');
}

var _exit;

function _load_exit() {
  return _exit = require('../exit');
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.default = function (program) {
  program.command('publish [project-dir]').alias('p').description('Publishes your project to exp.host').option('-q, --quiet', 'Suppress verbose output from the React Native packager.').option('-s, --send-to [dest]', 'A phone number or email address to send a link to').option('-c, --clear', 'Clear the React Native packager cache')
  // TODO(anp) set a default for this dynamically based on whether we're inside a container?
  .option('--max-workers [num]', 'Maximum number of tasks to allow Metro to spawn.').option('--release-channel <release channel>', "The release channel to publish to. Default is 'default'.", 'default').asyncActionProjectDir(action, true);
};
