// expo-cli/build/commands/client/clientBuildApi.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getExperienceName = exports.createClientBuildRequest = undefined;

var _regenerator;

function _load_regenerator() {
  return _regenerator = _interopRequireDefault(require('babel-runtime/regenerator'));
}

var _stringify;

function _load_stringify() {
  return _stringify = _interopRequireDefault(require('babel-runtime/core-js/json/stringify'));
}

var _asyncToGenerator2;

function _load_asyncToGenerator() {
  return _asyncToGenerator2 = _interopRequireDefault(require('babel-runtime/helpers/asyncToGenerator'));
}

var createClientBuildRequest = function () {
  var _ref = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee(_ref2) {
    var _ref2$user = _ref2.user,
        user = _ref2$user === undefined ? null : _ref2$user,
        context = _ref2.context,
        distributionCert = _ref2.distributionCert,
        pushKey = _ref2.pushKey,
        udids = _ref2.udids,
        addUdid = _ref2.addUdid,
        email = _ref2.email;
    return (_regenerator || _load_regenerator()).default.wrap(function _callee$(_context) {
      while (1) {
        switch (_context.prev = _context.next) {
          case 0:
            _context.next = 2;
            return (_xdl || _load_xdl()).ApiV2.clientForUser(user).postAsync('client-build/create-ios-request', {
              appleTeamId: context.team.id,
              appleTeamName: context.team.name,
              addUdid: addUdid,
              bundleIdentifier: context.bundleIdentifier,
              email: email,
              credentials: {
                apnsKeyP8: pushKey.apnsKeyP8,
                apnsKeyId: pushKey.apnsKeyId,
                certP12: distributionCert.certP12,
                certPassword: distributionCert.certPassword,
                teamId: context.team.id,
                appleSession: context.fastlaneSession,
                udidsString: (0, (_stringify || _load_stringify()).default)(udids)
              }
            });

          case 2:
            return _context.abrupt('return', _context.sent);

          case 3:
          case 'end':
            return _context.stop();
        }
      }
    }, _callee, this);
  }));

  return function createClientBuildRequest(_x) {
    return _ref.apply(this, arguments);
  };
}();

var getExperienceName = function () {
  var _ref3 = (0, (_asyncToGenerator2 || _load_asyncToGenerator()).default)( /*#__PURE__*/(_regenerator || _load_regenerator()).default.mark(function _callee2(_ref4) {
    var _ref4$user = _ref4.user,
        user = _ref4$user === undefined ? null : _ref4$user,
        appleTeamId = _ref4.appleTeamId;

    var _ref5, experienceName;

    return (_regenerator || _load_regenerator()).default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            _context2.next = 2;
            return (_xdl || _load_xdl()).ApiV2.clientForUser(user).postAsync('client-build/experience-name', {
              appleTeamId: appleTeamId
            });

          case 2:
            _ref5 = _context2.sent;
            experienceName = _ref5.experienceName;
            return _context2.abrupt('return', experienceName);

          case 5:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function getExperienceName(_x2) {
    return _ref3.apply(this, arguments);
  };
}();

var _xdl;

function _load_xdl() {
  return _xdl = require('xdl');
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

exports.createClientBuildRequest = createClientBuildRequest;
exports.getExperienceName = getExperienceName;
